ni_measurementlink_service/_internal/parameter/_message.py

```diff
@@ -1,30 +1,32 @@
 import struct
-from typing import Any, Dict
+from typing import Any, Callable, Dict, Optional, Tuple, Type, TypeVar
 
 from google.protobuf.internal import encoder, wire_format
 from google.protobuf.message import Message
 
 from ni_measurementlink_service._internal.parameter._serializer_types import (
     Decoder,
-    Encoder,
     Key,
     NewDefault,
+    WriteFunction,
 )
 
 
-def _message_encoder_constructor(field_index: int, is_repeated: bool, is_packed: bool) -> Encoder:
+def _message_encoder_constructor(
+    field_index: int, is_repeated: bool, is_packed: bool
+) -> Callable[[WriteFunction, Message, bool], int]:
     """Mimics google.protobuf.internal.MessageEncoder.
 
-    We cannot use MessageEncoder because it is for the pure-Python implementation
-    of messages. Our messages are upb messages, so we need to call 'SerializeToString'
+    This function was forked in order to call SerializeToString instead of _InternalSerialize.
 
-    See EncodeField:
-    https://github.com/protocolbuffers/protobuf/blob/0b817d46d4ca1977d3dccf2442aeee3c9e98e3a1/python/google/protobuf/internal/encoder.py#L765C15-L765C15
+    _InternalSerialize is only defined for the pure-Python protobuf implementation. Our child
+    messages (like DoubleXYData) are defined in .proto files, so they use whichever protobuf
+    implementation that google.protobuf.internal.api_implementation chooses (usually upb).
     """
     tag = encoder.TagBytes(field_index, wire_format.WIRETYPE_LENGTH_DELIMITED)
     encode_varint = _varint_encoder()
 
-    def _encode_message(write, value, deterministic):
+    def _encode_message(write: WriteFunction, value: Message, deterministic: bool):
         write(tag)
         bytes = value.SerializeToString()
         encode_varint(write, len(bytes), deterministic)
```

Comment:
This is missing the return type hint, so it's implicitly `Any`. The `Encoder` typedef suggests that it returns an `int`, so you should return the value of `write(bytes)`.

Protobuf seems to return an int for scalar writes and None for repeated writes, so maybe the correct return type is `Optional[int]`, but we don't have to worry about that in this case.